include/spirv-tools/optimizer.hpp

```diff
@@ -888,12 +888,31 @@ Optimizer::PassToken CreateAmdExtToKhrPass();
 Optimizer::PassToken CreateInterpolateFixupPass();
 
 // Removes unused components from composite input variables. Current
-// implementation just removes trailing unused components from input arrays.
-// The pass performs best after maximizing dead code removal. A subsequent dead
-// code elimination pass would be beneficial in removing newly unused component
-// types.
+// implementation just removes trailing unused components from input arrays
+// and structs. The pass performs best after maximizing dead code removal.
+// A subsequent dead code elimination pass would be beneficial in removing
+// newly unused component types.
+//
+// WARNING: This pass can only be safely applied standalone to vertex shaders
```

Comment:
Not a fan of this. I understand that the optimizer run() function now only takes one module, and thus this logic must be moved the the user, but could we add a "PipelineOptimizer" for such passes? It would overlay the Optimizer, and make sure those passes are correctly applied to all stages?

Is there cases where one wouldn't want to apply all those passes on some parts of the pipeline? (like you say here, running it only standalone on the vertex stage?)

Reply:
> Not a fan of this. I understand that the optimizer run() function now only takes one module, and thus this logic must be moved the the user, but could we add a "PipelineOptimizer" for such passes? It would overlay the Optimizer, and make sure those passes are correctly applied to all stages?

Yeah, I too would love to one day have a Pipeline Optimizer in spirv-tools. Think of these passes as building blocks and maybe even a blueprint toward that goal. By adding these building blocks now, we are making it that much easier for someone to create such a thing.

In the meantime, these passes will be used by our client to create a similar optimizer in their app. And someone has expressed an intent to investigate using these passes to optimize pipelines in an open-source Vulkan driver.

> Is there cases where one wouldn't want to apply all those passes on some parts of the pipeline? (like you say here, running it only standalone on the vertex stage?)

Yes. Actually, my next PR will be to add CreateEliminateDeadInputComponentsSafe(), which will only process vertex shaders. I will add --eliminate-dead-input-components-safe, and will retarget --eliminate-dead-input-components to use the safe version as well.